File: AdventOfCode2024/ruby/lib/day7/day_7.rb
class Day7

  INPUT_PATH = File.expand_path('../../../../inputs/day7/input.txt', __FILE__).freeze

  def self.puzzle1
    input = File.read(INPUT_PATH).split("\n")

    total_calibration_result = 0
    input.each do |line|
      target, numbers = line.split(':').map(&:strip)
      target = target.to_i
      numbers = numbers.split(/\s+/).map(&:to_i)
      total_calibration_result += target if can_make_value?(numbers, target)
    end

    total_calibration_result
  end

  def self.puzzle2
    input = File.read(INPUT_PATH).split("\n")

    total_calibration_result = 0
    input.each do |line|
      target, numbers = line.split(':').map(&:strip)
      target = target.to_i
      numbers = numbers.split(/\s+/).map(&:to_i)
      total_calibration_result += target if can_make_value_v2?(numbers, target)
    end

    total_calibration_result
  end

  def self.can_make_value?(numbers, target, current_value = nil, remaining_numbers = nil)
    # Initial setup on first call
    if remaining_numbers.nil?
      return false if numbers.empty?
      remaining_numbers = numbers.dup
      current_value = remaining_numbers.shift
    end

    # Base case: if no more numbers and current value equals target
    return current_value == target if remaining_numbers.empty?

    next_num = remaining_numbers[0]
    next_remaining = remaining_numbers[1..]

    # Try addition
    return true if can_make_value?(numbers, target, current_value + next_num, next_remaining)

    # Try multiplication
    return true if can_make_value?(numbers, target, current_value * next_num, next_remaining)

    # If neither operation works, return false
    false
  end

  def self.can_make_value_v2?(numbers, target, current_value = nil, remaining_numbers = nil)
    # Initial setup on first call
    if remaining_numbers.nil?
      return false if numbers.empty?
      remaining_numbers = numbers.dup
      current_value = remaining_numbers.shift
    end

    # Base case: if no more numbers and current value equals target
    return current_value == target if remaining_numbers.empty?

    next_num = remaining_numbers[0]
    next_remaining = remaining_numbers[1..]

    # Try addition
    return true if can_make_value_v2?(numbers, target, current_value + next_num, next_remaining)

    # Try multiplication
    return true if can_make_value_v2?(numbers, target, current_value * next_num, next_remaining)

    # Try concatenation
    return true if can_make_value_v2?(numbers, target, (current_value.to_s + next_num.to_s).to_i, next_remaining)

    # If neither operation works, return false
    false
  end
end


puts Day7.puzzle1
puts Day7.puzzle2

File: AdventOfCode2024/java/src/main/java/day7/Day7.java
package day7;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;

public class Day7 {
    public static final String INPUT_PATH = "../../../../inputs/day7/input.txt";

    public static long puzzle1() throws IOException {
        List<String> input = Files.readAllLines(Paths.get(INPUT_PATH));
        
        long totalCalibrationResult = 0;
        for (String line : input) {
            String[] parts = line.split(":");
            long target = Long.parseLong(parts[0].trim());
            long[] numbers = Arrays.stream(parts[1].trim().split("\\s+"))
                                .mapToLong(Long::parseLong)
                                .toArray();
            
            if (canMakeValue(numbers, target, null, null)) {
                totalCalibrationResult += target;
            }
        }
        
        return totalCalibrationResult;
    }

    public static long puzzle2() throws IOException {
        List<String> input = Files.readAllLines(Paths.get(INPUT_PATH));
        
        long totalCalibrationResult = 0;
        for (String line : input) {
            String[] parts = line.split(":");
            long target = Long.parseLong(parts[0].trim());
            long[] numbers = Arrays.stream(parts[1].trim().split("\\s+"))
                                .mapToLong(Long::parseLong)
                                .toArray();
            
            if (canMakeValueV2(numbers, target, null, null)) {
                totalCalibrationResult += target;
            }
        }
        
        return totalCalibrationResult;
    }

    private static boolean canMakeValue(long[] numbers, long target, Long currentValue, long[] remainingNumbers) {
        // Initial setup on first call
        if (remainingNumbers == null) {
            if (numbers.length == 0) return false;
            remainingNumbers = Arrays.copyOfRange(numbers, 1, numbers.length);
            currentValue = numbers[0];
        }

        // Base case: if no more numbers and current value equals target
        if (remainingNumbers.length == 0) {
            return currentValue == target;
        }

        long nextNum = remainingNumbers[0];
        long[] nextRemaining = Arrays.copyOfRange(remainingNumbers, 1, remainingNumbers.length);

        // Try addition
        if (canMakeValue(numbers, target, currentValue + nextNum, nextRemaining)) {
            return true;
        }

        // Try multiplication
        if (canMakeValue(numbers, target, currentValue * nextNum, nextRemaining)) {
            return true;
        }

        // If neither operation works, return false
        return false;
    }

    private static boolean canMakeValueV2(long[] numbers, long target, Long currentValue, long[] remainingNumbers) {
        // Initial setup on first call
        if (remainingNumbers == null) {
            if (numbers.length == 0) return false;
            remainingNumbers = Arrays.copyOfRange(numbers, 1, numbers.length);
            currentValue = numbers[0];
        }

        // Base case: if no more numbers and current value equals target
        if (remainingNumbers.length == 0) {
            return currentValue == target;
        }

        long nextNum = remainingNumbers[0];
        long[] nextRemaining = Arrays.copyOfRange(remainingNumbers, 1, remainingNumbers.length);

        // Try addition
        if (canMakeValueV2(numbers, target, currentValue + nextNum, nextRemaining)) {
            return true;
        }

        // Try multiplication
        if (canMakeValueV2(numbers, target, currentValue * nextNum, nextRemaining)) {
            return true;
        }

        // Try concatenation
        String concatenated = String.valueOf(currentValue) + String.valueOf(nextNum);
        if (canMakeValueV2(numbers, target, Long.parseLong(concatenated), nextRemaining)) {
            return true;
        }

        // If no operation works, return false
        return false;
    }

    public static void main(String[] args) throws IOException {
        System.out.println(puzzle1());
        System.out.println(puzzle2());
    }
} 